src/metrics/storage.go

```diff
@@ -0,0 +1,173 @@
+package metrics
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/fs"
+	"reflect"
+	"runtime"
+	"sync"
+
+	"github.com/boltdb/bolt"
+	log "github.com/sirupsen/logrus"
+)
+
+// PeerStoreStorage provides the necessary functions
+// to interact with peers data.
+// It is either saved in a Db or in Memory.
+type PeerStoreStorage interface {
+	Store(key string, value Peer)
+	Load(key string) (value Peer, ok bool)
+	Delete(key string)
+	Range(f func(key string, value Peer) bool)
+}
+
+// PeerStoreDb save the peer's data in a persistant Db.
+type PeerStoreDb struct {
+	db *BoltDB
+}
+
+// PeerStoreMemory save the peer's data in RAM.
+// Unless exported, data is lost after execution.
+type PeerStoreMemory struct {
+	m *sync.Map
+}
+
+func (p *PeerStoreDb) Store(key string, value Peer) {
+	value_marshalled, err := json.Marshal(value)
+	if err != nil {
+		log.Error(err)
+		return
+	}
+	p.db.Store([]byte(key), value_marshalled)
+}
+
+func (p *PeerStoreDb) Load(key string) (value Peer, ok bool) {
+	value_marshalled, ok := p.db.Load([]byte(key))
+	if !ok {
+		return Peer{}, false
+	}
+
+	err := json.Unmarshal(value_marshalled, &value)
+	if err != nil {
+		log.Error(err)
+		return Peer{}, false
+	}
+	return value, true
+}
+
+func (p *PeerStoreDb) Delete(key string) {
+	p.db.Delete([]byte(key))
+}
+
+func (p *PeerStoreDb) Range(f func(key string, value Peer) bool) {
+	p.db.Range(func(key, value []byte) bool {
+		var value_unmarshalled Peer
+		err := json.Unmarshal(value, &value_unmarshalled)
+		if err != nil {
+			log.Error(err)
+			return false
+		}
+		ok := f(string(key), value_unmarshalled)
+		return ok
+	})
+}
+
+func (p *PeerStoreMemory) Store(key string, value Peer) {
```

Comment:
What do you think about having only `Store(value Peer)` ? In our case the key is implicitly included in the `Peer` (`value.PeerId`). I think its dangerous having it like that because it can create inconsistencies.
Imagine that someone calls `Store("1", Peer{PeerId: "3",...}`. That would be inconsistent.